xbmc/video/dialogs/GUIDialogVideoBookmarks.cpp

```diff
@@ -161,38 +161,67 @@ bool CGUIDialogVideoBookmarks::OnAction(const CAction &action)
   return CGUIDialog::OnAction(action);
 }
 
+int CGUIDialogVideoBookmarks::ItemToBookmarkIndex(int item)
+{
+  if (item < 0 || item >= (int)m_vecItems->Size())
+    return -1;
+
+  std::shared_ptr<CFileItem> fileItem = m_vecItems->Get(item);
+
+  if (!fileItem->GetProperty("isbookmark").asBoolean(false))
+    return -1;
+
+  const int bookmarkIdx{fileItem->GetProperty("bookmark").asInteger32(-1)};
+  if (bookmarkIdx < 0 || bookmarkIdx >= m_bookmarks.size())
+  {
+    CLog::LogF(LOGERROR, "invalid bookmark index {} for {} bookmark(s)", bookmarkIdx,
+               m_bookmarks.size());
+    return -1;
+  }
+  return bookmarkIdx;
+}
+
 
 void CGUIDialogVideoBookmarks::OnPopupMenu(int item)
 {
-  if (item < 0 || item >= (int) m_bookmarks.size())
+  const int bookmarkIdx{ItemToBookmarkIndex(item)};
+  if (bookmarkIdx < 0)
     return;
 
+  const CBookmark& bm{m_bookmarks[bookmarkIdx]};
+
   // highlight the item
   (*m_vecItems)[item]->Select(true);
 
   CContextButtons choices;
-  choices.Add(1, (m_bookmarks[item].type == CBookmark::EPISODE ? 20405 : 20404)); // "Remove episode bookmark" or "Remove bookmark"
+  choices.Add(1, (bm.type == CBookmark::EPISODE
+                      ? 20405
+                      : 20404)); // "Remove episode bookmark" or "Remove bookmark"
 
-  int button = CGUIDialogContextMenu::ShowAndGetChoice(choices);
+  const int button{CGUIDialogContextMenu::ShowAndGetChoice(choices)};
 
   // unhighlight the item
   (*m_vecItems)[item]->Select(false);
 
   if (button == 1)
-    Delete(item);
+    Delete(bm);
 }
 
 void CGUIDialogVideoBookmarks::Delete(int item)
 {
-  if ( item>=0 && (unsigned)item < m_bookmarks.size() )
-  {
-    CVideoDatabase videoDatabase;
-    videoDatabase.Open();
-    const std::string path{g_application.CurrentFileItem().GetDynPath()};
-    videoDatabase.ClearBookMarkOfFile(path, m_bookmarks[item], m_bookmarks[item].type);
-    videoDatabase.Close();
-    CUtil::DeleteVideoDatabaseDirectoryCache();
-  }
+  const int bookmarkIdx{ItemToBookmarkIndex(item)};
+  if (bookmarkIdx >= 0)
+    Delete(m_bookmarks[bookmarkIdx]);
+}
+
+void CGUIDialogVideoBookmarks::Delete(const CBookmark& bm)
+{
+  CVideoDatabase videoDatabase;
+  videoDatabase.Open();
```

Comment:
You should handle videodb open not succeeding. At least that’s what we’re doing in most places on our code base. There must be a reason for this.

Reply:
That pattern is all over the place in that dialog. I'll add a commit for all of them.